src: explicitly register built-in modules

yhwang opened the pull request:
Previously, built-in modules are registered before main() via
`__attribute__((constructor))` mechanism in GCC and similiar
mechanism in MSVC. This causes some issues when node is built as
static library. Calling module registration function for built-in
modules in node::Init() helps to avoid the issues.

Refs: https://github.com/nodejs/node/pull/14986#issuecomment-332758206
Signed-off-by: Yihong Wang <yh.wang@ibm.com>

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

addaleax commented on src/node.h:
```diff
@@ -200,6 +200,7 @@ NODE_EXTERN extern bool force_fips_crypto;
 # endif
 #endif
 
+NODE_EXTERN void RegisterBuiltinModules();
```
Can you add a short doc comment for this method, in particular that embedders need to call it on startup?

I know a lot in this file isn’t very well-documented, but at least new code should come with something. :)

addaleax commented on src/node.h:
```diff
@@ -439,6 +440,48 @@ struct node_module {
+    V(inspector)                                                              \
+    V(crypto)                                                                 \
+    V(tls_wrap)
+
```
node_internals.h might be a better place?

addaleax commented on src/node.h:
```diff
@@ -458,10 +501,16 @@ extern "C" NODE_EXTERN void node_module_register(void* mod);
   NODE_CTOR_PREFIX void fn(void)
+#define NODE_C_CTOR_BUILTIN(fn)                                       \
+  void fn(void);                                                      \
+  void fn(void)
```
ditto re: node_internals.h

addaleax commented on src/node.h:
```diff
@@ -502,14 +551,40 @@ extern "C" NODE_EXTERN void node_module_register(void* mod);
     }                                                                 \
   }
 
+/* for nodejs built-in modules only */
```
This is a very good sign that this, too, would be nicer to have in `node_internals.h` :)

yhwang replied:
@addaleax thanks, actually @bnoordhuis had the same comments. I thought I moved it. I got 2-3 working servers, seems I didn't pick the right one to push the change. let me do it now. Thanks for the comments

yhwang commented:
I pushed an updated version. I should address the `node_internals.h` suggestions.

Fishrock123 commented on src/node.h:
```diff
@@ -508,8 +508,6 @@ extern "C" NODE_EXTERN void node_module_register(void* mod);
   NODE_MODULE_CONTEXT_AWARE_X(modname, regfunc, NULL, 0)
 
-#define NODE_MODULE_CONTEXT_AWARE_BUILTIN(modname, regfunc)           \
-  NODE_MODULE_CONTEXT_AWARE_X(modname, regfunc, NULL, NM_F_BUILTIN)   \
```
Probably a good idea but this is definitely `semver-major`.

addaleax replied:
@Fishrock123 What here is semver-major?

I’m on the fence myself but *if* it is semver-major, then that is the case because embedders need to make an extra function call – not because of the other changes?

addaleax commented on src/node_internals.h:
```diff
@@ -203,6 +297,14 @@ void SetupProcessObject(Environment* env,
+ * use the __attribute__((constructor)). Need to
+ * explicitly call the _register* functions.
+ */
+void RegisterBuiltinModules();
```
Sorry, I was unclear, but this is part of the embedder API so it should be in the public header – right?

yhwang replied:
Hi @addaleax and @gireeshpunathil,  actually, this change is for built-in modules. It only removes the `__attribute__((constructos))` for build-in modules. So, I moved this to the `node_internals.h` and it needs to be invoked inside the `node::Start()`. 

For third-party modules, they are built as shared library in most case and `__attribute__((constructor))` works properly in `dlopen()`. I left that part as-is. I was not aware that xIc compiler doesn't support it. If we want to remove the `__attribute__((constructor))` for third-party modules, we also need to change the module initialization process.

addaleax replied:
@yhwang Yes, but not all embedders use `node::Start(int argc, char** argv)`, right? So they need a definition of this function to be available.

bnoordhuis commented on src/node.cc:
```diff
@@ -127,6 +127,18 @@ typedef int mode_t;
+ * function for each built-in modules explicitly in
+ * node::RegisterBuiltinModules(). This is only forward declaration.
+ * The definitions are in each module's implementation when calling
+ * the NODE_MODULE_CONTEXT_AWARE_BUILTIN.
```
Tiny nit: can you use `//` comments?  C-style comments aren't exactly wrong but they look incongruous because we use C++-style comments everywhere else.

bnoordhuis commented on src/node_internals.h:
```diff
@@ -75,6 +75,100 @@ struct sockaddr;
+                                            priv,                             \
+                                            flags)                            \
+  static node::node_module _module =                                          \
+  {                                                                           \
```
Style nit: `{` should go on the previous line and since you're here, can you replace `NULL` with `nullptr`?

bnoordhuis commented on src/node_internals.h:
```diff
@@ -75,6 +75,100 @@ struct sockaddr;
+    V(uv,             MODREG_WITH_DEF)                                        \
+    V(inspector,      MODREG_WITH_DEF)                                        \
+    V(crypto,         OPENSSL_MODREG_DEF)                                     \
+    V(tls_wrap,       OPENSSL_MODREG_DEF)
```
This is not wrong but I'd approach it a little differently.
```c++
#define NODE_BUILTIN_STANDARD_MODULES(V) \
  V(async_wrap) \
  V(cares_wrap) \
  // ...

#if HAVE_OPENSSL
#define NODE_BUILTIN_OPENSSL_MODULES(V) V(crypto) V(tls_wrap)
#else
#define NODE_BUILTIN_OPENSSL_MODULES(V)
#endif

#if NODE_HAVE_I18N_SUPPORT
#define NODE_BUILTIN_ICU_MODULES(V) V(icu)
#else
#define NODE_BUILTIN_ICU_MODULES(V)
#endif

#define NODE_BUILTIN_MODULES(V) \
  NODE_BUILTIN_STANDARD_MODULES(V) \
  NODE_BUILTIN_OPENSSL_MODULES(V) \
  NODE_BUILTIN_ICU_MODULES(V)
```
It's arguably a little easier to comprehend because the V macro takes only one argument this way.

bnoordhuis commented on src/node_internals.h:
```diff
@@ -75,6 +75,100 @@ struct sockaddr;
+    NULL,                                                                     \
+    __FILE__,                                                                 \
+    NULL,                                                                     \
+    (node::addon_context_register_func) (regfunc),                            \
```
Something for a future cleanup PR: getting rid of this cast.

bnoordhuis commented on src/node_internals.h:
```diff
@@ -75,6 +75,100 @@ struct sockaddr;
+    NULL                                                                      \
+  };                                                                          \
+  NODE_C_CTOR_BUILTIN(_register_ ## modname) {                                \
+    node_module_register(&_module);                                           \
```
Another something for a future cleanup PR: export the struct and call `node_module_register()` directly in node.cc, instead of first going through `_register_foo()`.

Performance-wise it's unlikely to matter but it should make it a little easier to understand for future maintainers.

Also, I think you can get rid of `NODE_C_CTOR_BUILTIN` now, you don't need to special-case for Windows anymore.

bnoordhuis commented on src/node.cc:
```diff
@@ -127,6 +127,18 @@ typedef int mode_t;
+ * The definitions are in each module's implementation when calling
+ * the NODE_MODULE_CONTEXT_AWARE_BUILTIN.
+ */
+#define DECLAREFUN(modname, impl) void _register_##modname(void) impl
```
Tiny nit: `void _register_##modname()` (no void argument list, that's a C-ism) and perhaps you can call the macro simply V?  Also applies to line 4999.

yhwang commented:
@bnoordhuis I updated the commit to address all your comments.

@addaleax you're right. So I moved the invocation of `node::RegisterBuiltinModules()` into `node::Init()`. Is that okay? If you know some scenarios that embedders would still need the RegisterBuiltinModules(), I will move it to node.h.

bnoordhuis commented on src/node_internals.h:
```diff
@@ -75,6 +75,88 @@ struct sockaddr;
+    priv,                                                                     \
+    nullptr                                                                   \
+  };                                                                          \
+  void _register_ ## modname() {                                             \
```
`\` is not aligned with its surroundings.

bnoordhuis commented on src/node_internals.h:
```diff
@@ -325,7 +413,10 @@ class InternalCallbackScope {
-  NODE_MODULE_CONTEXT_AWARE_X(modname, regfunc, NULL, NM_F_INTERNAL)  \
+  NODE_MODULE_CONTEXT_AWARE_X_BUILTIN(modname,                        \
+                                      regfunc,                        \
+                                      NULL,                           \
```
Since you're here: nullptr

yhwang replied:
sure and done!

TimothyGu commented on src/node_internals.h:
```diff
@@ -75,6 +75,88 @@ struct sockaddr;
+    V(tty_wrap)                                                               \
+    V(udp_wrap)                                                               \
+    V(uv)                                                                     \
+    V(inspector)
```
nit: alphabetical order please :)

TimothyGu commented on test/cctest/node_module_reg.cc:
```diff
@@ -0,0 +1,28 @@
+void _register_fs_event_wrap() {}
+void _register_js_stream() {}
+void _register_module_wrap() {}
+void _register_config() {}
```
nit: alphabetical order

yhwang replied:
fixed.

yhwang commented:
Hi @Fishrock123 and @bnoordhuis I removed the change in `src/node.h` and all changes go to `src/node_internals` now. Can you review again? Thanks.

bnoordhuis commented on src/node_v8.cc:
```diff
@@ -20,6 +20,7 @@
 // USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 #include "node.h"
+#include "node_internals.h"
```
Is this include and the one in uv.cc necessary?

yhwang replied:
Yes. need to add `node_internals.h` for node_v8.cc and uv.cc

yhwang commented:
@bnoordhuis thanks for the review. would you remove the `semver-major`? label? I can do the `node.h` clean up in a separated PR.

mhdawson commented:
CI Run: https://ci.nodejs.org/job/node-test-pull-request/11135/

yhwang commented:
I found the failures seems related to this https://github.com/nodejs/node/issues/16210 and I updated the commit to sync up with upstream.

mhdawson commented:
New CI run: https://ci.nodejs.org/job/node-test-pull-request/11159/

yhwang commented:
the failures are:
- parallel/test-http2-compat-serverrequest-pipe # TODO : Fix flaky test
- parallel/test-http2-pipe # TODO : Fix flaky test
- parallel/test-async-wrap-uncaughtexception
- pi1-raspbian-wheezy

first three are flaky tests and I am not sure about the last one. there are 6 pi1 jobs, 1 passed and 5 failed.

yhwang commented:
Ping, Is there anything that I should do in order to get this to be landed?

yhwang commented:
fetch upstream and rebase my change. this needs a new CI, please.

gibfahn commented:
@yhwang this needs another rebase.

yhwang commented:
@gibfahn resolved the conflict now. please kick off a new CI, thanks.

gibfahn commented:
CI: https://ci.nodejs.org/job/node-test-commit/13981/

yhwang commented:
I think this is the CI: https://ci.nodejs.org/job/node-test-pull-request/11397/

gibfahn commented:
@yhwang correct, I linked to the `node-test-commit` job because it's more useful that linking to the `test-pull-request` job (you can directly see which platforms passed or failed).

<img width="614" alt="image" src="https://user-images.githubusercontent.com/15943089/32752796-7715f0a4-c8c2-11e7-9b78-f57f6169c9ae.png">

gibfahn commented:
Landed in https://github.com/nodejs/node/commit/8680bb9f1a0163cfbdc4443c1eb2b56c5e443616

MylesBorins commented:
This is not landing cleanly on v9.x, could you please manually backport?

yhwang commented:
@MylesBorins can you give me more information about this? or anything I can try to fix?

MylesBorins commented:
We build the 9.x release out of the `v9.x-staging` branch
when attempting to run `git cherry-pick 8680bb9`  to add this commit we are getting conflicts

specifically in `src/node_internals.h`. Digging a bit deeper it seems like this change relies on bits from https://github.com/nodejs/node/pull/16567

If it relies on the bits from that PR that are Semver-Major then we likely shouldn't land this in 9.x

/cc @addaleax and @jasnell who might be able to give more info

yhwang commented:
thanks for the information. I just check the #16567. It added bunch of `ERRNO_xxxx` macro.  the `8680bb9` shouldn't depends on those macro. I tried to do `git cherry-pick 8680bb9` locally in `v9.x-staging`. Yes, there is conflict. I solved it locally and I can built node without any issue. Let me attach the diff for you
[node_internals.h.diff.txt](https://github.com/nodejs/node/files/1549670/node_internals.h.diff.txt)

MylesBorins commented:
If you have a working version of the patch locally would you be able to
open a PR against v9.x-staging?

Please maintain all the original meta data in the commit message

On Dec 11, 2017 6:02 PM, "Yihong Wang" <notifications@github.com> wrote:

> thanks for the information. I just check the #16567
> <https://github.com/nodejs/node/pull/16567>. It added bunch of ERRNO_xxxx
> macro. the 8680bb9 shouldn't depends on those macro. I tried to do git
> cherry-pick 8680bb9 locally in v9.x-staging. Yes, there is conflict. I
> solved it locally and I can built node without any issue. Let me attach the
> diff for you
> node_internals.h.diff.txt
> <https://github.com/nodejs/node/files/1549670/node_internals.h.diff.txt>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/16565#issuecomment-350887342>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAecV4GC_p6GmOKUsft0D4B2Vue8ZMqGks5s_bR1gaJpZM4QJn0b>
> .
>

yhwang commented:
I created a PR for it and please kick off a CI for it. Thanks
